usr.bin/fold/fold.1
.\"	$NetBSD: fold.1,v 1.14.2.1 2012/05/23 10:08:24 yamt Exp $
.\"
.\" Copyright (c) 1980, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)fold.1	8.1 (Berkeley) 6/6/93
.\"
.Dd May 12, 2012
.Dt FOLD 1
.Os
.Sh NAME
.Nm fold
.Nd "fold long lines for finite width output device"
.Sh SYNOPSIS
.Nm
.Op Fl bs
.Op Fl w Ar width
.Op Ar
.Sh DESCRIPTION
.Nm
is a filter which folds the contents of the specified files,
or the standard input if no files are specified,
breaking the lines to have a maximum of 80 characters.
.Pp
The options are as follows:
.Bl -tag -width XwXwidthXX
.It Fl b
Count
.Ar width
in bytes rather than column positions.
.It Fl s
Fold line after the last blank character within the first
.Ar width
column positions (or bytes).
If a blank character does not exist within the width, then
a longer line will still be split at the width.
.It Fl w Ar width
Specifies
.Ar width
to use as a line width, instead of the default 80 characters.
.El
.Sh ENVIRONMENT
.Bl -tag -width indent
.It Ev LC_CTYPE
.El
.Sh EXIT STATUS
.Ex -std
.Sh SEE ALSO
.Xr expand 1 ,
.Xr fmt 1
.Sh STANDARDS
The
.Nm
utility conforms to
.St -p1003.1-2008 .
.Sh HISTORY
The
.Nm
utility appeared in
.Bx 1 .
.Sh BUGS
If underlining is present it may be messed up by folding.
.Pp
.Ar Width
should be a multiple of 8 if tabs are present, or the tabs should
be expanded using
.Xr expand 1
before using
.Nm .


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision: 1.1703 $>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------


Changes from NetBSD 6.0 to NetBSD 7.0:
	window(1): Removed from the base system in favor of tmux(1).  Use the
		pkgsrc/misc/window package from now on.  [jmmv 20120215]
	pkg_add(1): imported 20120221 version from pkgsrc [wiz 20120221]
	file(1): upgraded to 5.11 [christos 20120222]
	evbmips: Add support for Lemote Yeeloong Notebook. [nonaka 20120302]
	dbsym(8): Add option -p [bsh 20120319]
	dhcpcd(8): Import dhcpcd-5.5.5 [roy 20120323]
	ipf: upgrade to 5.1.1 with new build framework, no module and
		test support yet [christos 20120323]
	urtwn(4): Add a driver for Realtek RTL8188CU/RTL8192CU USB IEEE
		802.11b/g/n wireless network device, ported from OpenBSD.
		[nonaka 20120325]
	pcc: upgrade to 1.1.0.DEVEL 20120325 [plunky 20120325]
	openresolv(8): Import openresolv-3.5.1 [roy 20120327]
	dhcpcd(8): Import dhcpcd-5.5.6 [roy 20120328]
	apple: Recognize the apple disk partitioning scheme, and auto-discover
		wedges. Fix iPod disk mounting, by adding quirk.
		[christos 20120306]
	fstab: Add NAME=wedgename for fs_spec and routines to parse it. Now
		wedges can be used in /etc/fstab, even in dynamic environments
		where disks come and go and wedges get renumbered.
		[christos 20120306]
	pdisk(8): Make it work for 64 bit hosts, and make it MI.
		[christos 20120306]
	amiga: Add clockport(4) layer and a driver for A1200 on-board clockport
		a1k2cp(4). [rkujawa 20120417]
	kernel: add mpii(4), a driver for LSI Logic Fusion-MPT Message Passing
		Interface II SAS controllers.  [bouyer 20120419]
	kernel: Add support for 256 CPUs on amd64. [rmind 20120420]
	OpenSSH: Imported 6.0 [christos 20120501]
	bzip2: Imported 1.0.6 [wiz 20120507]
	amiga: Rework X-Surf driver as xsurf(4). Add support for clockport(4)
		on top of xsurf(4). [rkujawa 20120515]
	backtrace(3): Add <execinfo.h> and the backtrace family of functions
		[christos 20120526]


usr.bin/fold/fold.c
/*	$NetBSD: fold.c,v 1.16 2009/07/21 01:35:02 ahoka Exp $	*/

/*-
 * Copyright (c) 1990, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Kevin Ruddy.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#ifndef lint
__COPYRIGHT("@(#) Copyright (c) 1990, 1993\
 The Regents of the University of California.  All rights reserved.");
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)fold.c	8.1 (Berkeley) 6/6/93";
#endif
__RCSID("$NetBSD: fold.c,v 1.16 2009/07/21 01:35:02 ahoka Exp $");
#endif /* not lint */

#include <limits.h>
#include <locale.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <wchar.h>
#include <err.h>

#define	DEFLINEWIDTH	80

	int	main(int, char **);
static	void	fold(int);
static	int	new_column_position(int, wint_t);
static	void	usage(void);

int count_bytes = 0;
int split_words = 0;

int
main(int argc, char **argv)
{
	int ch;
	int width;
	char *p;

	setlocale(LC_CTYPE, "");
	setprogname(argv[0]);

	width = -1;
	while ((ch = getopt(argc, argv, "0123456789bsw:")) != -1)
		switch (ch) {
		case 'b':
			count_bytes = 1;
			break;
		case 's':
			split_words = 1;
			break;
		case 'w':
			if ((width = atoi(optarg)) <= 0)
				errx(1, "illegal width value");
			break;
		case '0': case '1': case '2': case '3': case '4':
		case '5': case '6': case '7': case '8': case '9':
			if (width == -1) {
				p = argv[optind - 1];
				if (p[0] == '-' && p[1] == ch && !p[2])
					width = atoi(++p);
				else
					width = atoi(argv[optind] + 1);
			}
			break;
		default:
			usage();
		}
	argv += optind;
	argc -= optind;

	if (width == -1)
		width = DEFLINEWIDTH;

	if (!*argv)
		fold(width);
	else for (; *argv; ++argv)
		if (!freopen(*argv, "r", stdin)) {
			err (1, "%s", *argv);
			/* NOTREACHED */
		} else
			fold(width);
	exit(0);
}

/*
 * Fold the contents of standard input to fit within WIDTH columns
 * (or bytes) and write to standard output.
 *
 * If split_words is set, split the line at the last space character
 * on the line.  This flag necessitates storing the line in a buffer
 * until the current column > width, or a newline or EOF is read.
 *
 * The buffer can grow larger than WIDTH due to backspaces and carriage
 * returns embedded in the input stream.
 */
static void
fold(int width)
{
	static wchar_t *buf = NULL;
	wchar_t *nbuf;
	static int   buf_max = 0;
	wint_t ch;
	int col, indx, i;

	col = indx = 0;
	while ((ch = getwchar()) != WEOF) {
		if (ch == L'\n') {
			if (indx != 0) {
				for (i = 0; i < indx; i++)
					putwchar(buf[i]);
			}
			putwchar(L'\n');
			col = indx = 0;
			continue;
		}

		col = new_column_position (col, ch);
		if (col > width) {
			int last_space;

#ifdef __GNUC__
			last_space = 0;	/* XXX gcc */
#endif
			if (split_words) {
				for (i = 0, last_space = -1; i < indx; i++)
					if (buf[i] == L' ')
						last_space = i;
			}

			if (split_words && last_space != -1) {
				for (i = 0; i < last_space; i++)
					putwchar(buf[i]);

				/* increase last_space here, so we skip trailing whitespace */
				last_space++;
				wmemmove (buf, buf+last_space, indx-last_space);

				indx -= last_space;
				col = 0;
				for (i = 0; i < indx; i++) {
					col = new_column_position (col, buf[i]);
				}
			} else {
				for (i = 0; i < indx; i++)
					putwchar(buf[i]);
				col = indx = 0;
			}
			putwchar('\n');

			/* calculate the column position for the next line. */
			col = new_column_position (col, ch);
		}

		if (indx + 1 > buf_max) {
			/* Allocate buffer in LINE_MAX increments */
			if ((nbuf = realloc (buf, buf_max + 2048)) == NULL) {
				err (1, "realloc");
				/* NOTREACHED */
			}
			buf = nbuf;
			buf_max += 2048;
		}
		buf[indx++] = ch;
	}

	if (indx != 0) {
		for (i = 0; i < indx; i++)
			putwchar(buf[i]);
	}
}

/*
 * calculate the column position 
 */
static int
new_column_position (int col, wint_t ch)
{
	int w;
	
	if (!count_bytes) {
		switch (ch) {
		case L'\b':
			if (col > 0)
				--col;
			break;
		case L'\r':
			col = 0;
			break;
		case L'\t':
			col = (col + 8) & ~7;
			break;
		default:
			w = wcwidth(ch);
			if (w > 0)
				col += w;
			break;
		}
	} else {
		char dummy[MB_LEN_MAX];
		
		/* XXX: we assume stateless encoding */
		col += wcrtomb(dummy, ch, NULL);
	}

	return col;
}

static void
usage(void)
{
	(void)fprintf(stderr,
	    "usage: %s [-bs] [-w width] [file ...]\n", getprogname());
	exit(1);
}

